server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -216,11 +217,24 @@ protected void newResponseAsync(
 
     protected abstract NodeResponse newNodeResponse(StreamInput in, DiscoveryNode node) throws IOException;
 
+    /**
+     * Implements the request recipient logic.
+     * If access to the channel listener is needed, override {@link #nodeOperationAsync(TransportRequest, ChannelActionListener, Task)}.
+     */
     protected abstract NodeResponse nodeOperation(NodeRequest request, Task task);
 
     /**
-     * resolve node ids to concrete nodes of the incoming request
-     **/
+     * This method can be overridden if a subclass needs to asynchronously respond to the node request via a listener.
+     * The default implementation is to fall through to {@link #nodeOperation}.
+     */
+    protected void nodeOperationAsync(NodeRequest request, ChannelActionListener<NodeResponse> listener, Task task) {
```

Comment:
nit: could we make this just an `ActionListener` instead of a `ChannelActionListener` (and ideally make it the last argument)? There should be no need for implementations to know they're getting a `ChannelActionListener`, this should just be a regular async method.

Reply:
sgtm, done 👍

Thanks for taking a look :)